### ISSUE TITLE
Refactor engine to support rendering to multiple windows from a single Isolate

### ISSUE THREAD

**Issue by ColdPaleLight** [flutter/flutter#96531]
We intend to experiment with supporting rendering to multiple windows from a single Isolate.

According to the design document flutter.dev/go/desktop-multi-window-support, we need to support a single engine with multiple surfaces.

However, this is not easy. The codes in `Shell`, `PlatformView`, `Engine`, `Animator`, and `Rasterizer`  etc. in the engine are all designed based on a single engine with single surface. 

Let's take `Rasterizer` as an example. If we want to let `Rasterizer` support single engine with multiple surfaces
1.  The member variables `surface_`, `last_layer_tree_`, `resubmitted_layer_tree_` are originally designed for single engine with single surface, so their type must be changed to `unordered_map`
2.  Methods such as `Rasterizer::Draw`, `Rasterizer::DoDraw`, `Rasterizer::DrawToSurface`, etc. all need to add a parameter `window_id` to find the specified `surface`, `last_layer_tree`, `resubmitted_layer_tree` .
``` diff 

RasterStatus Rasterizer::Draw(
+ int64_t window_id,
  std::unique_ptr<FrameTimingsRecorder> frame_timings_recorder,
  std::shared_ptr<Pipeline<flutter::LayerTree>> pipeline,
  LayerTreeDiscardCallback discard_callback) {
```

`Shell`, `Animator`, `PlatformView` also need to be changed in the above way. Therefore, there will be many types of variables need to be changed to `unordered_map`, and many methods to add a parameter named `window_id`.

I don't know if such a change makes sense, it looks a little weird. but it looks like you'd have to make similar changes if you want to support single engine with multiple surfaces. Of course, this is just the tip of the iceberg. There are still many things that need to be changed in the engine to support multiple windows.

/cc @Hixie @chinmaygarde @gaaclarke @dnfield @knopp

**Comment by knopp** [comment]
Perhaps it might make more sense to have one `Rasterizer` instance per view/surface?

**Comment by ColdPaleLight** [comment]
> Perhaps it might make more sense to have one `Rasterizer` instance per view/surface?

If so, the relationship between Shell and Rasterizer will change from 1v1 to 1vN, and for the same reason, PlatformView will probably need to make the same change.

I'm not sure if it makes sense, it breaks the 1v1 relationship between the shell and its components.

**Comment by dnfield** [comment]
Perhaps there'd be some class or struct that holds a surface for a window and its associated last layer tree, resubmitted layer tree, etc.

I think it would help though to have a diagram of where things would go from `window.render` on through to rasterization so we could more easily discuss the architecture.

**Comment by knopp** [comment]
We will need to break the 1v1 to 1vN relationship somewhere.

**Comment by ColdPaleLight** [comment]
> I think it would help though to have a diagram of where things would go from window.render on through to rasterization so we could more easily discuss the architecture.

Great idea, I will do this.

**Comment by ColdPaleLight** [comment]
> We will need to break the 1v1 relationship somewhere.

Yes, but I guess we may need to keep the 1v1 relationship between Shell and its components (PlatformView, Engine, Rasterizer, ShellIOManager) , and make 1v1 to 1vN changes inside the components (for example, Rasterizer will have multiple surfaces, PlatformView will be renamed to PlatformBinding and then there are multiple PlatformWindow inside, etc.)

**Comment by ColdPaleLight** [comment]
@knopp 
After a bit of trying, I think you're making a lot of sense. We should make a `Rasterizer` correspond to a Window/Surface. Handling 1vN relationships inside the Rasterizer makes everything complicated.

**Comment by zanderso** [comment]
Routing to @goderbauer to decide whether this is adequately tracked by this issue or a different issue.

**Comment by wanjm** [comment]
where can i see the progress on this ? is there any git prototyping this ?

**Comment by dkwingsmt** [comment]
We're tracking our progress at https://github.com/orgs/flutter/projects/39. There's an MVP app at https://github.com/goderbauer/mvp .

**Comment by JensvandeWiel** [comment]
I would like to help, is there anything i can do. I am quite new to flutter but i need this feature so i want to help.

**Comment by dkwingsmt** [comment]
> I would like to help, is there anything i can do. I am quite new to flutter but i need this feature so i want to help.

Thank you for your offer, although we've at the late stage of our phase 1, which will enable plugins  to manage windows through native code. What you can do will be to help beta test the API once it's close to ready, to review our design, and to create a plugin for it when it's available.

**Comment by JensvandeWiel** [comment]
Alright,  ill be sure to do that!

**Comment by wanjm** [comment]
how many phase are there in total?

**Comment by dkwingsmt** [comment]
We're currently splitting the work into 2 phases. You can find an introduction at the beginning of https://docs.google.com/document/d/1U_ciMTHJgDrnw8kIo9sGKhiKFOycWRmz1yVVI_osCXg/edit

**Comment by loic-sharma** [comment]
We're still actively working on this. This work item contains more granular info on the remaining work: https://github.com/flutter/flutter/issues/142845
